### ISSUE TITLE
[RFC] A high-level GSPMD API in PT/XLA (based on `xs.mark_sharding`)

### ISSUE THREAD

**Issue by ronghanghu** [pytorch/xla#3755]
## 🚀 [RFC] A high-level GSPMD API in PT/XLA (based on `xs.mark_sharding`)

This RFC proposes a high-level API for GSPMD through a wrapper class and a partitioning rule function, based on `xs.mark_sharding`.

## Motivation

[GSPMD](https://arxiv.org/abs/2105.04663) is a powerful approach for model sharding and parallelization, and is enabled in PyTorch/XLA by the `xs.mark_sharding` API in https://github.com/pytorch/xla/pull/3476 and https://github.com/pytorch/xla/pull/3684. However, there are a few limitations when directly using this API to build SPMD programs:

* To build SPMD (e.g. on a transformer model) directly using this API, one would have to edit the PyTorch model and insert `xs.mark_sharding` into the initialization and forward code of each layer in the model. This forces the model implementation to be aware of the SPMD sharding strategies and makes the model code device-dependent and sharding-dependent, incompatible with the native PyTorch, and hard to switch to other sharding strategies (e.g. Megatron->Optimus) or other parallelism APIs (e.g. the [FSDP](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py) wrapper).
* The requirement of adding `xs.mark_sharding` to a model implementation (e.g. its forward code) makes it hard/impossible to run SPMD in those cases where the users cannot easily edit the model implementation code (e.g. when a model comes from external libraries) and largely limits its application. **For example, suppose one wants to apply SPMD to a torchvision model or a Hugging Face transformer, it is infeasible to submit a pull request to these libraries to add `xs.mark_sharding` into their model implementations.**
* SPMD applications often require annotating a lot of tensors, sometimes on those submodules several levels below the base model (e.g. every transformer MLP `nn.Linear` layer in a T5 encoder-decoder model for Megatron-style sharding), so it is inconvenient and error-prong to directly build SPMD programs on top of `xs.mark_sharding`.

Ideally, one would like to be able to take any existing PyTorch models (e.g. BERT from Hugging Face), and apply a specific sharding strategy to it (e.g. Megatron), without changing the model code (in Hugging Face library code in this case).

In this RFC, we advocate for a high-level SPMD API in PyTorch/XLA (built upon `xs.mark_sharding`) without requesting the user to rewrite their entire model code to be aware of the SPMD sharding partitions. **Specifically, we propose to parallelize a model with SPMD by first building the base model and then wrapping it with a wrapper class `SPDMParallel`, similar to how DDP or FSDP is applied to an existing model.**

The proposed `SPDMParallel` class takes as input

1. a given PyTorch `module` (the base model, an `nn.Module` instance), and
2. a user-specified `sharding_rule_func` to define the sharding rules, and
3. a `mesh_shape` tuple to define the TPU mesh.

Since those tensors we want to shard in SPMD are usually either the parameters or the input/output tensors of a submodule (e.g. an `nn.Linear`) in the whole network, the `SPDMParallel` class will recursively go down to all the submodules in the wrapped model and apply the sharding rules to the parameters, inputs, and outputs of each submodule. The sharded HLO graph will be built when forward (and backward) is called on the `SPDMParallel` class.

In this way, the base model code doesn't need to be changed when using SPMD -- it will just be wrapped and partitioned in a post-hoc manner, so all the current model implementations from the abundant existing PyTorch libraries (e.g. timm, torchvision, Hugging Face) can be directly used for SPMD parallelism without rewriting these libraries.

## Proposed Implementation

The proposed implementation (prototype) of the high-level SPMD API consists of a `SPDMParallel` class that recursively applies a `sharding_rule_func` to the submodules of an input `module`, which can be based on their names and the submodule instances themselves. A prototype is as follows:

```python3
class SPDMParallel(nn.Module):
  """recursively apply a `sharding_rule_func` to submodules of an input `module`."""

  def __init__(self, module, sharding_rule_func, mesh_shape):
      super().__init__()
  
      # apply SPMD sharding rule to the base model
      sharded_module = self.apply_sharding(module, sharding_rule_func, mesh_shape)
      self.module = sharded_module
      self.sharding_rule_func = sharding_rule_func
      self.mesh_shape = mesh_shape

  def apply_sharding(self, module, sharding_rule_func, mesh_shape):
      sharded_module = deepcopy(module)  # make a copy so that we don't change the input `module`

      # recursively apply the sharding rule to all the submodules
      for name, m in module.named_modules():
          sharding_rule_func(name, m, mesh_shape)

      return sharded_module

  def forward(self, *args, **kwargs):
      return self.module(*args, **kwargs)
```

And an example sharding rule function (`sharding_rule_func` above) will apply a specific SPMD sharding annotation to an `nn.Module`'s parameters, inputs, and outputs. For example, the following sharding rule function can be used to apply the Megatron sharding to those MLP layers in a [timm Vision Transformer (ViT)](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py).

```python3
def vit_megatron_rule(name: str, submodule: nn.Module, mesh_shape: Tuple[Union[int, None]]):
    
    # assuming `mesh_shape` is in (data-parallel, model-parallel) format like in T5X

    if name.endswith("mlp.fc1"):
      assert isinstance(submodule, nn.Linear)
      # shard the 1st MLP layer's weight param (mlp_dim, hidden_size)
      submodule.weight = xs.mark_sharding(submodule.weight, mesh_shape, (1, None))
      # shard the 1st MLP layer's bias param (mlp_dim,)
      submodule.bias = xs.mark_sharding(submodule.bias, mesh_shape, (1,))
      
      # shard the 1st MLP layer's output (batch_size, seq_length, mlp_dim) by patching its forward
      # TODO (come up with a more readable way to patch module input and output)
      def _forward_sharded_output(m, x):
          return xs.mark_sharding(m._orig_forward(x), mesh_shape, (None, 1))
      submodule._orig_forward = submodule.forward
      submodule.forward = MethodType(_forward_sharded_output, submodule)

    elif name.endswith("mlp.fc2"):
      assert isinstance(submodule, nn.Linear)

      # shard the weight (hidden_size, mlp_size) in the 2nd MLP layer
      submodule.weight = xs.mark_sharding(submodule.weight, mesh_shape, (None, 1))
```

We could also provide a few easier ways to build `sharding_rule_func` or to use it in `SPDMParallel`.

## Alternatives

The implementation above requires a `sharding_rule_func` (that takes a submodule object and its name) to decide how to decorate its parameters, inputs, or return values using the `xs.mark_sharding` API. While this should be sufficient to implement nearly all SPMD use cases, it is hard to later inspect the sharding annotations in a `SPDMParallel` instance, such as printing a list of sharded tensors annotated by `xs.mark_sharding` and their partitioning details. It relies on the user to keep track of what is sharded by `sharding_rule_func`.

An alternative way to implement this `SPDMParallel` class is to have a more principled way to define the sharding rule. Rather than having an arbitrary function to do anything, one can have a name-based (string-based) sharding rule similar to the [logical axis names](https://github.com/google-research/t5x/blob/main/docs/usage/partitioning.md#canonical-logical-axis-names) in T5X. Under this name-based sharding rule definition, a sharding rule consists of the following:

1) a function to extract those tensors to be sharded (param, input tensor, return values) from a submodule and map their tensor axes to logical axis names (such as `(batch, mlp_size)`).

2) a user-specified mapping rule (e.g. a list) to map a logical axis to a TPU mesh axis, [similar to those in T5X](https://github.com/google-research/t5x/blob/main/docs/usage/partitioning.md#data-and-model-parallel-with-replicated-activations).

An orthogonal and complementary way to simplify the sharding rule is to use [Named Tensors API](https://pytorch.org/docs/stable/named_tensor.html) to give each tensor axis a name (a string). If the users need to implement both their base model (to be wrapped by `SPDMParallel`) as well as their `sharding_rule_func`, then it would be easier for them to use named tensors in their base model implementation and refer to those axis names in their `sharding_rule_func` implementation. However, this approach cannot be applied to existing models (e.g. those in torchvision or Hugging Face) that don't use the named tensors, so named tensors should not be a requirement in the `SPDMParallel` class.

## Additional context

We are happy to work on a prototype implementation on our end and first try it out in our internal use cases. We and submit a PR once we have a mature implementation.

cc: @yeounoh @JackCaoG @miladm @ultrons

**Comment by yeounoh** [comment]
Thank you @ronghanghu , let's build this :)

**Comment by yeounoh** [comment]
Hi @ronghanghu , this may or may not be relevant. The above examples seem correct, but for what it's worth;  mark_sharding returns an XlaShardedTensor, but it's just a wrapper around the original tensor -- mark_sharding still passes the tensor to the backend with the sharding spec. This means that we don't need to replace the module layer/params with the output or mark_sharding. Here is an example:
```
device = xm.xla_device()
# load the predefined model
model = get_model_property('model_fn')().to(device)
# mark sharding
for name, layer in model.named_modules():
  if 'conv' in name:
    partition = int(xm.xrt_world_size() / 2)
    xs.mark_sharding(layer.weight, (1, 1, partition, partition), (0, 1, 2, 3))
```

**Comment by ronghanghu** [comment]
> mark_sharding returns an XlaShardedTensor, but it's just a wrapper around the original tensor -- mark_sharding still passes the tensor to the backend with the sharding spec. This means that we don't need to replace the module layer/params with the output or mark_sharding.

This is great to know and makes a lot of things easier. 

If we can do 
```
y = x * 2
xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
# Case 1: using the existing tensor `x` instead of the output of `xs.mark_sharding`
z = y + 1
```
instead of 
```
y = x * 2
y_shard = xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
# Case 2: using the output of `xs.mark_sharding`
z = y_shard + 1
```
then it makes the entire life much easier since we can still have the same python object id, etc. This allows us to e.g. first construct the entire model and the optimizer, and then just mark a few of its outputs.

I wonder how would Case 1 work exactly. Is `y` still a regular `torch.Tensor` or a `XLAShardedTensor`? And should `xs.mark_sharding` return `None` if this is just an in-place annotation?

**Comment by ronghanghu** [comment]
> mark_sharding returns an XlaShardedTensor, but it's just a wrapper around the original tensor -- mark_sharding still passes the tensor to the backend with the sharding spec. This means that we don't need to replace the module layer/params with the output or mark_sharding.

@yeounoh This is great to know and makes a lot of things easier :) Thanks for the update!

If we can do 
```
y = x * 2
xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
# Case 1: using the existing tensor `x` instead of the output of `xs.mark_sharding`
z = y + 1
```
instead of 
```
y = x * 2
y_shard = xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
# Case 2: using the output of `xs.mark_sharding`
z = y_shard + 1
```
then it makes the entire life much easier since we can still have the same python object id, etc. This allows us to e.g. first construct the entire model and the optimizer, and then just mark a few of its outputs.

I wonder how would Case 1 work exactly. Is `y` still a regular `torch.Tensor` or a `XLAShardedTensor`? And should `xs.mark_sharding` return `None` if this is just an in-place annotation?

**Comment by yeounoh** [comment]
All three scenarios are possible, since the annotation is first attached to the original tensor and XLAShardedTensor wraps around it. XLAShardedTensor may be needed more explicitly if a user wants to use its APIs.

For the second case, 
```
y = x * 2
z = y + 1
# Case 2: like Case 1 but running `mark_sharding` at the end
xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
```
If user executed (e.g., print) `z` before mark_sharding, then it would have been computed without sharding; the executions after mark_sharding will be partitioned.

**Comment by ronghanghu** [comment]
@yeounoh I see, this is great to hear and should make the SPMD execution much simpler. We probably won't even need a wrapper class at all, and perhaps a few annotation utilities will be enough.

Another question is -- are the annotations marked by `xs.mark_sharding` persistent through `xm.mark_step`, or do we need to mark them again every iteration? The application scenario is whether we need to call `xs.mark_sharding` on a parameter `layer.weight` in every training iteration, or do we just need to do it only once at the beginning of the training?

**Comment by JackCaoG** [comment]
`It will persist through xm.mark_step` 

I am wondering how is this done. I thought sharding_spec is a field in the `IR`, upon `mark_step` we will clear all IR and replace them with device_data. Through I guess in this case device data will be in multiple device so it is still sharded.

**Comment by yeounoh** [comment]
> 

@JackCaoG yea it's persisted through the xla_data, which is PjRtShardedData. I am thinking about using the IR as a single source of truth for the shaing annotation, and resetting the IR will check & preserve the sharding annotation if it exists. Let me follow up with you offline.

**Comment by yeounoh** [comment]
@ronghanghu I had a discussion with @JackCaoG , and we've decided to block mark_sharding to take effect if it's called after. It's always forward-looking and thus, appear eager.
```
y = x * 2
z = y + 1   # execute unpartitioned
# Case 2: like Case 1 but running `mark_sharding` at the end
xs.mark_sharding(y, (1, 1, partition, partition), (0, 1, 2, 3))
z = y + 1  # execute partitioned
```

**Comment by ronghanghu** [comment]
@yeounoh I see, sounds good and thanks for letting me know!

**Comment by JackCaoG** [comment]
for the real model, it should looks more like
```
input = torch.tensor()..
model = nn.linear()
xs.mark_sharding(model.parameters()[0], spec)           # not a expert, not sure if we need to do `.data`
xs.mark_sharding(input, spec)
res = model(input)
```

**Comment by ronghanghu** [comment]
@JackCaoG Cool, this usage looks good to me :)

**Comment by ronghanghu** [comment]
@yeounoh Another relevant usage issue (that I also discussed with Jack today) is how to initialize model parameters in SPMD under a multi-host setting.

In the `xs.mark_sharding(model.parameters()[0], spec)` example above, suppose we do the vanilla data parallelism where the model parameters should be replicated across the TPU mesh. When we have multiple hosts (e.g. v3-32,) each host might initialize its parameter to different values if we follow the typical way of first initializing the parameters on the CPU and then casting them to XLA devices (unless we explicitly enforce the same random seed, which people typically don't do). So we might end up marking the model parameters as replicated when they are actually not replicated (same shape but different values).

PyTorch DDP deals with this issue with a broadcast-based synchronization when initializing the DDP wrapper. I'm thinking we can implement a similar sync in our `GSPMDParallel` wrapper class or a sync function to do a parameter broadcasting in a multi-host SPMD setting, before actually starting the training.

**Comment by yeounoh** [comment]
> @yeounoh Another relevant usage issue (that I also discussed with Jack today) is how to initialize model parameters under SPMD in a multi-host setting.
> 
> 
> 
> In the `xs.mark_sharding(model.parameters()[0], spec)` example above, suppose we do the vanilla data parallelism where the model parameters should be replicated across the TPU mesh. When we have multiple hosts (e.g. v3-32), each host might initialize its parameter to different values if we follow the typical way of first initializing the parameters on the CPU and then casting them to XLA devices (unless we explicitly enforce the same torch random seed everywhere, which people typically don't do, and is often not possible e.g. because we need different data augmentation on each host). So we might end up marking the model parameters as replicated when they are actually not replicated (same shape but different values).
> 
> 
> 
> PyTorch DDP deals with this issue with a broadcast-based synchronization when initializing the DDP wrapper. I'm thinking we can implement a similar sync in our `GSPMDParallel` wrapper class or a sync function to do a parameter broadcasting in a multi-host SPMD setting, before actually starting the training.
> 
> 

I see, I was hoping to handle this in PyToch/XLA layer, so that the user doesn't need to do anything but treat the multi-host case the same as a single replica but just with more cores. Let's resume the discussion when we start pod testing.

**Comment by ColdCodeCool** [comment]
hi all, is there any update for this RFC? Has it been merged into the code base?

**Comment by alanwaketan** [comment]
> hi all, is there any update for this RFC? Has it been merged into the code base?

No, it has not been merged to the code base yet. We are looking into integrating this API into distributed_module listed above.
